#Thinkful Data Science Course
##Unit 4: Predicting the Future; 
##Lesson 5: Clustering
##Cluster Analysis

####Summary:

Now we're going to apply K-Means clustering to a data set from the UN on economic and demographic data for different countries. We'll also use some well-known techniques to find the right K to use.

Load the UN countries dataset.
Determine how many rows are in the dataset. Determine the number of non-null values present in each column.
Based on the number of non-null values, which columns do you think will be the best to try to cluster on?
Determine the data type of each column.
How many counries are present in the dataset?
We're going to see how lifeMale, lifeFemale and infantMortality cluster according to GDPperCapita, keeping in mind that we don't know in advance how many clusters there will be.
Before we cluster the dataset, we need to determine the ideal number of clusters. For this, we need numpy and scipy.

We're going to explore 1 to 10 clusters. Define a range K from 1 to 10.
Apply scipy.cluster.vq.kmeans 1 to 10 and determine the cluster centroids.
Calculate the distance between each point and each cluster centroid.
Determine the closest centroid for each point.
Calculate the average within-cluster sum of squares for each centroid.
When we cluster, we are effectively trying to minimize the Euclidian distance between a cluster centroid and the points in that cluster. Practically, we define our clusters based on what will minimize what's called the within-cluster sum of squares (WCSS) (since sum of squares is the squared Euclidian distance).

Plot the number of clusters against the average within-cluster sum of squares. At what number of clusters does the curve seem to "level out"?

Now we're going to the cluster the dataset. From the elbow plot we see that the largest drops in average within-cluster sum of squares occur from 1 to 2 and from 2 to 3 clusters. After that the drops are much smaller and decreasing. We pick 3 as the best number of clusters for this dataset.

Cluster the UN dataset using k-means clustering with 3 clusters.
Plot the results. What patterns stand out to you? What conclusions can you draw?

In [17]:
import pandas as pd
import numpy as np
from scipy.cluster.vq import kmeans, vq,  whiten
from pylab import plot,show
import collections

####Load the UN countries dataset.

In [4]:
UNdf = pd.read_csv('/Users/molliepettit/Desktop/Data Science/Thinkful/Projects/Unit4/un.csv')

In [7]:
UNdf

country   region   tfr  contraception  educationMale  \
0             Afghanistan     Asia  6.90            NaN            NaN   
1                 Albania   Europe  2.60            NaN            NaN   
2                 Algeria   Africa  3.81             52           11.1   
3          American.Samoa     Asia   NaN            NaN            NaN   
4                 Andorra   Europe   NaN            NaN            NaN   
5                  Angola   Africa  6.69            NaN            NaN   
6                 Antigua  America   NaN             53            NaN   
7               Argentina  America  2.62            NaN            NaN   
8                 Armenia   Europe  1.70             22            NaN   
9               Australia  Oceania  1.89             76           16.3   
10                Austria   Europe  1.42             71           14.4   
11             Azerbaijan     Asia  2.30             17            NaN   
12                Bahamas  America  1.95             62           12.1   
13                Bahrain     Asia  2.97             53           12.6   
14             Bangladesh     Asia  3.14             49            NaN   
15               Barbados  America  1.73             55            NaN   
16                Belarus   Europe  1.40             50            NaN   
17                Belgium   Europe  1.62             79           15.6   
18                 Belize  America  3.66             47           10.6   
19                  Benin   Africa  5.83             16            NaN   
20                 Bhutan     Asia  5.89             19            NaN   
21                Bolivia  America  4.36             45            NaN   
22                 Bosnia   Europe  1.40            NaN            NaN   
23               Botswana   Africa  4.45             33           10.5   
24                 Brazil  America  2.17             74            NaN   
25                 Brunei     Asia  2.70            NaN           11.8   
26               Bulgaria   Europe  1.45            NaN           11.8   
27           Burkina.Faso   Africa  6.57              8            3.3   
28                Burundi   Africa  6.28              9            5.1   
29               Cambodia     Asia  4.50            NaN            NaN   
..                    ...      ...   ...            ...            ...   
177             Swaziland   Africa  4.46             20           11.5   
178                Sweden   Europe  1.80             78           13.9   
179           Switzerland   Europe  1.46             71           14.5   
180                 Syria     Asia  4.00             36            9.8   
181            Tajikistan     Asia  3.93             21            NaN   
182              Tanzania   Africa  5.48             18            NaN   
183              Thailand     Asia  1.74             74            NaN   
184                  Togo   Africa  6.08             12            NaN   
185                 Tonga  Oceania  4.02             74            NaN   
186   Trinidad.and.Tobago  America  2.10             53           10.1   
187               Tunisia   Africa  2.92             60            NaN   
188                Turkey     Asia  2.50             63           10.6   
189          Turkmenistan     Asia  3.58             20            NaN   
190                Tuvalu  Oceania   NaN            NaN            NaN   
191                Uganda   Africa  7.10             15            NaN   
192               Ukraine   Europe  1.38             23            NaN   
193  United.Arab.Emirates     Asia  3.46            NaN            9.8   
194        United.Kingdom   Europe  1.72             82           16.1   
195         United.States  America  1.96             71           15.4   
196               Uruguay  America  2.25            NaN            NaN   
197            Uzbekistan     Asia  3.48             56            NaN   
198               Vanuatu  Oceania  4.36             15            NaN   
199             Venezu

####Determine how many rows are in the dataset.

In [6]:
UNdf.shape

(207, 14)

There are 207 rows and 14 columns.

####Determine the number of non-null values present in each column. 

In [8]:
UNdf.count()

country                   207
region                    207
tfr                       197
contraception             144
educationMale              76
educationFemale            76
lifeMale                  196
lifeFemale                196
infantMortality           201
GDPperCapita              197
economicActivityMale      165
economicActivityFemale    165
illiteracyMale            160
illiteracyFemale          160
dtype: int64

####Based on the number of non-null values, which columns do you think will be the best to try to cluster on

"country" and "region" have the most non-null values, so we should try to cluster on these columns.

####Determine the data type of each column. 

In [9]:
UNdf.dtypes

country                    object
region                     object
tfr                       float64
contraception             float64
educationMale             float64
educationFemale           float64
lifeMale                  float64
lifeFemale                float64
infantMortality           float64
GDPperCapita              float64
economicActivityMale      float64
economicActivityFemale    float64
illiteracyMale            float64
illiteracyFemale          float64
dtype: object

####How many counries are present in the dataset?

In [21]:
len(pd.unique(UNdf['country'].values.ravel()))

207

Entered code above to make sure to count only unique listings. Turns out they were all unique, so the number is still 207.

####We're going to see how lifeMale, lifeFemale and infantMortality cluster according 
to GDPperCapita, keeping in mind that we don't know in advance how many clusters there will be. Before we cluster the dataset, we need to determine the ideal number of clusters. For this, we need numpy and scipy, which were imported at top.

We're going to explore 1 to 10 clusters. Define a range K from 1 to 10. Apply scipy.cluster.vq.kmeans 1 to 10 and determine the cluster centroids. Calculate the distance between each point and each cluster centroid. Determine the closest centroid for each point. Calculate the average within-cluster sum of squares for each centroid. When we cluster, we are effectively trying to minimize the Euclidian distance between a cluster centroid and the points in that cluster. Practically, we define our clusters based on what will minimize what's called the within-cluster sum of squares (WCSS) (since sum of squares is the squared Euclidian distance).

Retrieving appropriate columns:

In [23]:
lifemale = UNdf['lifeMale']
gdp = UNdf['GDPperCapita']
lifefemale = UNdf['lifeFemale']
infantmortality = UNdf['infantMortality']

Replacing NaN values with zero:

In [24]:
lifemale[np.isnan(lifemale)]  = 0
gdp[np.isnan(gdp)] = 0
lifefemale[np.isnan(lifefemale)] = 0
infantmortality[np.isnan(infantmortality)] = 0

/Users/molliepettit/anaconda/lib/python3.4/site-packages/IPython/kernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
/Users/molliepettit/anaconda/lib/python3.4/site-packages/IPython/kernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from IPython.kernel.zmq import kernelapp as app
/Users/molliepettit/anaconda/lib/python3.4/site-packages/IPython/kernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  app.

Will create a data frame with the two columns:

In [27]:
d1 = {'gdp': gdp, 'lifemale': lifemale}
df1 = pd.DataFrame(d1)
cluster1 = df1.values

In [28]:
df1

gdp  lifemale
0     2848      45.0
1      863      68.0
2     1531      67.5
3        0      68.0
4        0       0.0
5      355      44.9
6     6966       0.0
7     8055      69.6
8      354      67.2
9    20046      75.4
10   29006      73.7
11     321      66.5
12   12545      70.5
13    9073      71.1
14     280      58.1
15    7173      73.6
16     994      64.4
17   26582      73.9
18    2569      73.4
19     391      52.4
20     166      51.6
21     909      59.8
22     271      70.5
23    3640      48.9
24    4510      63.4
25   16683      73.4
26    1518      67.8
27     165      45.1
28     205      45.5
29     130      52.6
..     ...       ...
177   1389      57.7
178  26253      76.2
179  42416      75.3
180   3573      66.7
181    122      64.2
182    139      50.0
183   2896      66.3
184    322      48.8
185   1787      67.0
186   4083      71.5
187   2030      68.4
188   2814      66.5
189    321      61.2
190      0       0.0
191    305      40.4
192    694      63.6
193  17690      73.9
194  18913      74.5
195  26037      73.4
196   5602      69.6
197    435      64.3
198   1289      65.5
199   3496      70.0
200    270      64.9
201      0       0.0
202      0      59.8
203    732      57.4
204   1487      69.8
205    382      42.2
206    786      47.6

[207 rows x 2 columns]